staging/src/k8s.io/controller-manager/config/v1alpha1/types.go

```diff
@@ -0,0 +1,45 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package config
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// LeaderMigrationConfiguration provides versioned configuration for all migrating leader locks.
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+type LeaderMigrationConfiguration struct {
```

Comment:
I'm not sure I understand the rationale for plumbing this as a config file, which I guess is the plan? It seems like it will be easy to accidentally get it different between controller managers.

Did you think about maybe using a configmap to make sure that all controller binaries can agree on this?

Sorry for not asking this in the KEP :)

@cheftako @liggitt

Reply:
Consuming as a file lets you feed in config from practically any source, including a configmap.

Tying directly to an API-observed configmap either assumes that you'll consume the config from the same cluster you're managing (which isn't always desired) or requires adding a second API client config path.

cc @deads2k for opinions on config via file vs API-fetched configmap.

Reply:
So the plan is a config file. I prefer the config file as its local to the control plane and safe from something like being accidentally deleted be a customer etc. Also we can ensure that the config files exist as part of all relevant version deployments as they are in the image, in either direction (upgrade or downgrade). Ensuring that the resource exists seems more error prone, especially in the downgrade case. Also the clean up of the file seems more natural. The 1.21 to 1.22 migration file exists in all 1.21 and 1.22 control plane releases. For the resource I have to worry about when its safe to delete (and re-add) the resource.

Reply:
Also as a resource this becomes and extension point for customers to disable built in controllers. Not sure that is something I want to "accidentally" implement.